hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/index/RunIndexActionExecutor.java

```diff
@@ -0,0 +1,384 @@
+    metaClient.getTableConfig().setValue(HoodieTableConfig.TABLE_METADATA_PARTITIONS.key(), String.join(",", completedPartitions));
+    HoodieTableConfig.update(metaClient.getFs(), new Path(metaClient.getMetaPath()), metaClient.getTableConfig().getProps());
+  }
+
+  /**
+   * Indexing check runs for instants that completed after the base instant (in the index plan).
+   * It will check if these later instants have logged updates to metadata table or not.
+   * If not, then it will do the update. If a later instant is inflight, it will wait until it is completed or the task times out.
+   */
+  class IndexingCatchupTask implements Runnable {
+
+    private final HoodieTableMetadataWriter metadataWriter;
+    private final List<HoodieInstant> instantsToIndex;
+    private final Set<String> metadataCompletedInstants;
+    private final HoodieTableMetaClient metaClient;
+    private final HoodieTableMetaClient metadataMetaClient;
+
+    IndexingCatchupTask(HoodieTableMetadataWriter metadataWriter,
+                        List<HoodieInstant> instantsToIndex,
+                        Set<String> metadataCompletedInstants,
+                        HoodieTableMetaClient metaClient,
+                        HoodieTableMetaClient metadataMetaClient) {
+      this.metadataWriter = metadataWriter;
+      this.instantsToIndex = instantsToIndex;
+      this.metadataCompletedInstants = metadataCompletedInstants;
+      this.metaClient = metaClient;
+      this.metadataMetaClient = metadataMetaClient;
+    }
+
+    @Override
+    public void run() {
+      for (HoodieInstant instant : instantsToIndex) {
+        // metadata index already updated for this instant
+        if (!metadataCompletedInstants.isEmpty() && metadataCompletedInstants.contains(instant.getTimestamp())) {
+          currentIndexedInstant = instant.getTimestamp();
+          continue;
+        }
+        while (!instant.isCompleted()) {
+          try {
+            LOG.warn("instant not completed, reloading timeline " + instant);
+            // reload timeline and fetch instant details again wait until timeout
+            String instantTime = instant.getTimestamp();
+            Option<HoodieInstant> currentInstant = metaClient.reloadActiveTimeline()
+                .filterCompletedInstants().filter(i -> i.getTimestamp().equals(instantTime)).firstInstant();
+            instant = currentInstant.orElse(instant);
+            // so that timeline is not reloaded very frequently
+            Thread.sleep(TIMELINE_RELOAD_INTERVAL_MILLIS);
+          } catch (InterruptedException e) {
+            throw new HoodieIndexException(String.format("Thread interrupted while running indexing check for instant: %s", instant), e);
+          }
+        }
+        // if instant completed, ensure that there was metadata commit, else update metadata for this completed instant
```

Comment:
So this is just so that any race causing the inflight to miss this is handled?

Reply:
yes right..